apis/apiextensions/v1alpha1/infrastructuredefinition_types.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+)
+
+// InfrastructureDefinitionSpec specifies the desired state of the definition.
+type InfrastructureDefinitionSpec struct {
+
+	// ConnectionSecretKeys is the list of keys that will be exposed to the end
+	// user of the defined kind.
+	ConnectionSecretKeys []string `json:"connectionSecretKeys,omitempty"`
+
+	// TODO(muvaf): As of this writing, CustomResourceDefinitionSpec structs
+	// do not have proper jsontags(validation substruct in particular), hence
+	// cannot be used as the type of the template. For now, it's RawExtension,
+	// but we definitely need to revisit so that proper validation is in place.
+
+	// CRDSpecTemplate is the base CRD template. The final CRD will have additional
+	// fields to the base template to accommodate Crossplane machinery.
+	CRDSpecTemplate runtime.RawExtension `json:"crdSpecTemplate,omitempty"`
```

Comment:
I wonder whether we should include the actual CRD spec struct here, or a stripped down port thereof?

Reply:
Here is an issue opened in controller-tools https://github.com/kubernetes-sigs/controller-tools/issues/291#issuecomment-554449809

Reply:
Oh and also they've got a few fields with type `*float64` which controller-tools outright rejects to work with.

Reply:
Alright I think I understand why they use custom OpenAPI functions for those structs. It essentially creates a recursive loop and the yaml never ends. I discovered that when I made a shallow copy of `CustomResourceDefinitionSpec` with a custom `JSONSchemaProps` whose field types that refer to `JSONSchemaProps` have proper jsontags. This causes infinite recursive loops and controller-tools just leaves the type as `{}` and this causes error in api-server even though it's not rejected when you `apply` the CRD yaml. See errors below:
```
  - lastTransitionTime: "2020-04-08T15:35:26Z"
    message: '[spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[validation].properties[openAPIV3Schema].properties[additionalItems].properties[schema].type:
      Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[validation].properties[openAPIV3Schema].properties[additionalProperties].properties[schema].type:
      Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[validation].properties[openAPIV3Schema].properties[allOf].items.type:
      Required value: must not be empty for specified array items, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[validation].properties[openAPIV3Schema].properties[anyOf].items.type:
      Required value: must not be empty for specified array items, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[validation].properties[openAPIV3Schema].properties[definitions].additionalProperties.type:
      Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[validation].properties[openAPIV3Schema].properties[dependencies].additionalProperties.properties[schema].type:
      Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[validation].properties[openAPIV3Schema].properties[items].properties[jsonSchemas].items.type:
      Required value: must not be empty for specified array items, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[validation].properties[openAPIV3Schema].properties[items].properties[schema].type:
      Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[validation].properties[openAPIV3Schema].properties[not].type:
      Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[validation].properties[openAPIV3Schema].properties[oneOf].items.type:
      Required value: must not be empty for specified array items, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[validation].properties[openAPIV3Schema].properties[patternProperties].additionalProperties.type:
      Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[validation].properties[openAPIV3Schema].properties[properties].additionalProperties.type:
      Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[versions].items.properties[schema].properties[openAPIV3Schema].properties[additionalItems].properties[schema].type:
      Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[versions].items.properties[schema].properties[openAPIV3Schema].properties[additionalProperties].properties[schema].type:
      Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[versions].items.properties[schema].properties[openAPIV3Schema].properties[allOf].items.type:
      Required value: must not be empty for specified array items, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[versions].items.properties[schema].properties[openAPIV3Schema].properties[anyOf].items.type:
      Required value: must not be empty for specified array items, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[versions].items.properties[schema].properties[openAPIV3Schema].properties[definitions].additionalProperties.type:
      Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[versions].items.properties[schema].properties[openAPIV3Schema].properties[dependencies].additionalProperties.properties[schema].type:
      Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[versions].items.properties[schema].properties[openAPIV3Schema].properties[items].properties[jsonSchemas].items.type:
      Required value: must not be empty for specified array items, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[versions].items.properties[schema].properties[openAPIV3Schema].properties[items].properties[schema].type:
      Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[versions].items.properties[schema].properties[openAPIV3Schema].properties[not].type:
      Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[versions].items.properties[schema].properties[openAPIV3Schema].properties[oneOf].items.type:
      Required value: must not be empty for specified array items, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[versions].items.properties[schema].properties[openAPIV3Schema].properties[patternProperties].additionalProperties.type:
      Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[spec].properties[crdSpecTemplate].properties[versions].items.properties[schema].properties[openAPIV3Schema].properties[properties].additionalProperties.type:
      Required value: must not be empty for specified object fields]'
```

So, we can make the spec `RawExtension` directly or make only `JSONSchemaProps` type. Like the following:
```go
// CustomResourceDefinitionSpec is a shallow copy of actual v1beta1.CustomResourceDefinitionSpec
type CustomResourceDefinitionSpec struct {
	...
	// validation describes the schema used for validation and pruning of the custom resource.
	// If present, this validation schema is used to validate all versions.
	// Top-level and per-version schemas are mutually exclusive.
	// +optional
	Validation *CustomResourceValidation `json:"validation,omitempty" protobuf:"bytes,5,opt,name=validation"`
        ...
	// versions is the list of all API versions of the defined custom resource.
	// Optional if `version` is specified.
	// The name of the first item in the `versions` list must match the `version` field if `version` and `versions` are both specified.
	// Version names are used to compute the order in which served versions are listed in API discovery.
	// If the version string is "kube-like", it will sort above non "kube-like" version strings, which are ordered
	// lexicographically. "Kube-like" versions start with a "v", then are followed by a number (the major version),
	// then optionally the string "alpha" or "beta" and another number (the minor version). These are sorted first
	// by GA > beta > alpha (where GA is a version with no suffix such as beta or alpha), and then by comparing
	// major version, then minor version. An example sorted list of versions:
	// v10, v2, v1, v11beta2, v10beta3, v3beta1, v12alpha1, v11alpha2, foo1, foo10.
	// +optional
	Versions []CustomResourceDefinitionVersion `json:"versions,omitempty" protobuf:"bytes,7,rep,name=versions"`
	...
}

// CustomResourceDefinitionVersion describes a version for CRD.
type CustomResourceDefinitionVersion struct {
        ...
	// schema describes the schema used for validation and pruning of this version of the custom resource.
	// Top-level and per-version schemas are mutually exclusive.
	// Per-version schemas must not all be set to identical values (top-level validation schema should be used instead).
	// +optional
	Schema *CustomResourceValidation `json:"schema,omitempty" protobuf:"bytes,4,opt,name=schema"`
	...
}

// CustomResourceValidation is a list of validation methods for CustomResources.
type CustomResourceValidation struct {
	// openAPIV3Schema is the OpenAPI v3 schema to use for validation and pruning.
	// +optional
	OpenAPIV3Schema runtime.RawExtension `json:"openAPIV3Schema,omitempty" protobuf:"bytes,1,opt,name=openAPIV3Schema"`
}
```

This way we get some validation but the conversion is trickier: both field assignments and conversion from `[]byte` need to happen. But it should be fine.